# Numerical features:

+ amount_tsh: the amount of water: it contains 0s as invaild values. nonlinear transformation does not help to remove the 0 values
+ gps_height: the values look ok no invaild values (it not possible to tell if the 0s are invaild values)
+ longitude: there are some 0 values as invalid values and they should be replaced
+ latitude: No sign of invalid values
+ population: there are some zeros as invalid values

# Categorical features:

+ region_code: there are some outliers with log transformation the problem can be solved
+ district_code: there are some outliers but the log transform it can solved
+ funder: 3635 missing values, a lot of categories
+ installer: 3655 missing values, a lot of categories
+ wpt_name: a lot of categories there are some None values
+ payment: limited categories no sign for missing values
+ payment_type: limited categories no sign for missing values (there are some 0s)
+ water_quality:limited categories no sign for missing values (there are some 0s)
+ quality_group:limited categories no sign for missing values (there are some 0s)
+ source:limited categories no sign for missing values (there are some 0s)
+ source_type:limited categories no sign for missing values (there are some 0s) (high correlated with soruce)
+ source_class: alot of 0s-->unknown
+ waterpoint_type:limited categories no sign for missing values (there are some 0s)
+ waterpoint_type_group:limited categories no sign for missing values (there are some 0s)
+ basin:limited categories no sign for missing values
+ subvillage: 371 missing values, a lot of categories there are no sign of missing values
+ region: limited categories no sign for missing values
+ lga:a lot of categories there are no sign of missing values
+ ward:a lot of categories there are no sign of missing values
+ public_meeting: 3334 missing values:limited categories there are some signs for missing values (there are some 0s and -1)
+ recorded_by: one value no variation
+ scheme_management: 3877 missing values limited categories
+ scheme_name: 28166 missing values lot of categories
+ permit:3056 missing value two categories
+ extraction_type:limited categories no sign for missing values
+ extraction_type_group:limited categories no sign for missing values
+ extraction_type_class:limited categories no sign for missing values
+ management:limited categories no sign for missing values
+ management_group:limited categories no sign for missing values
+ quantity:limited categories no sign for missing values
+ quantity_group:limited categories no sign for missing values

# Labels:
There are three labels status_group that should be predicted:

+ functional
+ non functional
+ functional needs repair.

## Importing relevant libraries and packages

In [75]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")
%matplotlib inline
import warnings
warnings.filterwarnings("ignore")

### Reading the data

The data that is goingto be used in this project is from 2 datasets. One with all the feature information and another that contains the target or the value to be predicted

In [76]:
#imports the Tanzania water training data set

filepath = "data/tanzania_water_training.csv"
details_df = pd.read_csv(filepath)
pd.set_option('display.max_columns', None)
details_df.head(10)

id  amount_tsh date_recorded               funder  gps_height  \
0  69572      6000.0    2011-03-14                Roman        1390   
1   8776         0.0    2013-03-06              Grumeti        1399   
2  34310        25.0    2013-02-25         Lottery Club         686   
3  67743         0.0    2013-01-28               Unicef         263   
4  19728         0.0    2011-07-13          Action In A           0   
5   9944        20.0    2011-03-13  Mkinga Distric Coun           0   
6  19816         0.0    2012-10-01                 Dwsp           0   
7  54551         0.0    2012-10-09                Rwssp           0   
8  53934         0.0    2012-11-03             Wateraid           0   
9  46144         0.0    2011-08-03          Isingiro Ho           0   

      installer  longitude   latitude              wpt_name  num_private  \
0         Roman  34.938093  -9.856322                  none            0   
1       GRUMETI  34.698766  -2.147466              Zahanati            0   
2  World vision  37.460664  -3.821329           Kwa Mahundi            0   
3        UNICEF  38.486161 -11.155298  Zahanati Ya Nanyumbu            0   
4       Artisan  31.130847  -1.825359               Shuleni            0   
5           DWE  39.172796  -4.765587                Tajiri            0   
6          DWSP  33.362410  -3.766365            Kwa Ngomho            0   
7           DWE  32.620617  -4.226198          Tushirikiane            0   
8     Water Aid  32.711100  -5.146712     Kwa Ramadhan Musa            0   
9       Artisan  30.626991  -1.257051               Kwapeto            0   

                     basin       subvillage     region  region_code  \
0               Lake Nyasa         Mnyusi B     Iringa           11   
1            Lake Victoria          Nyamara       Mara           20   
2                  Pangani          Majengo    Manyara           21   
3  Ruvuma / Southern Coast       Mahakamani     Mtwara           90   
4            Lake Victoria       Kyanyamisa     Kagera           18   
5                  Pangani      Moa/Mwereme      Tanga            4   
6                 Internal    Ishinabulandi  Shinyanga           17   
7          Lake Tanganyika  Nyawishi Center  Shinyanga           17   
8          Lake Tanganyika       Imalauduki     Tabora           14   
9            Lake Victoria         Mkonomre     Kagera           18   

   district_code              lga        ward  population public_meeting  \
0              5           Ludewa    Mundindi         109           True   
1              2        Serengeti       Natta         280            NaN   
2              4        Simanjiro     Ngorika         250           True   
3             63         Nanyumbu    Nanyumbu          58           True   
4              1          Karagwe  Nyakasimbi           0           True   
5              8           Mkinga         Moa           1           True   
6              3  Shinyanga Rural      Samuye           0           True   
7              3           Kahama      Chambo           0           True   
8              6     Tabora Urban    Itetemia           0           True   
9              1          Karagwe      Kaisho           0           True   

               recorded_by scheme_management                  scheme_name  \
0  GeoData Consultants Ltd               VWC                        Roman   
1  GeoData Consultants Ltd             Other                          NaN   
2  GeoData Consultants Ltd               VWC  Nyumba ya mungu pipe scheme   
3  GeoData Consultants Ltd               VWC                          NaN   
4  GeoData Consultants Ltd               NaN                          NaN   
5  GeoData Consultants Ltd               VWC                    Zingibali   
6  GeoData Consultants Ltd               VWC                          NaN   
7  GeoData Consultants Ltd               NaN                          NaN   
8  GeoData Consultants Ltd               VWC                          NaN  

In [77]:
details_df.shape

(59400, 40)

The data set contains 59400 records and 40 columns. That is alot and some of the columns will not make it to our analysis and modelling.

Below are the 40 column names and the datatypes of the column data

In [78]:
details_df.dtypes

id                         int64
amount_tsh               float64
date_recorded             object
funder                    object
gps_height                 int64
installer                 object
longitude                float64
latitude                 float64
wpt_name                  object
num_private                int64
basin                     object
subvillage                object
region                    object
region_code                int64
district_code              int64
lga                       object
ward                      object
population                 int64
public_meeting            object
recorded_by               object
scheme_management         object
scheme_name               object
permit                    object
construction_year          int64
extraction_type           object
extraction_type_group     object
extraction_type_class     object
management                object
management_group          object
payment                   object
payment_ty

In [79]:
#importing the target data set

conditions_df = pd.read_csv("data/training_set_labels.csv")
conditions_df.head()

id    status_group
0  69572      functional
1   8776      functional
2  34310      functional
3  67743  non functional
4  19728      functional

In [80]:
conditions_df.shape

(59400, 2)

In [81]:
#merging the 2 data sets to form 1 big one

all_df = details_df.merge(conditions_df[["id", "status_group"]], on="id")
all_df.head()

id  amount_tsh date_recorded        funder  gps_height     installer  \
0  69572      6000.0    2011-03-14         Roman        1390         Roman   
1   8776         0.0    2013-03-06       Grumeti        1399       GRUMETI   
2  34310        25.0    2013-02-25  Lottery Club         686  World vision   
3  67743         0.0    2013-01-28        Unicef         263        UNICEF   
4  19728         0.0    2011-07-13   Action In A           0       Artisan   

   longitude   latitude              wpt_name  num_private  \
0  34.938093  -9.856322                  none            0   
1  34.698766  -2.147466              Zahanati            0   
2  37.460664  -3.821329           Kwa Mahundi            0   
3  38.486161 -11.155298  Zahanati Ya Nanyumbu            0   
4  31.130847  -1.825359               Shuleni            0   

                     basin  subvillage   region  region_code  district_code  \
0               Lake Nyasa    Mnyusi B   Iringa           11              5   
1            Lake Victoria     Nyamara     Mara           20              2   
2                  Pangani     Majengo  Manyara           21              4   
3  Ruvuma / Southern Coast  Mahakamani   Mtwara           90             63   
4            Lake Victoria  Kyanyamisa   Kagera           18              1   

         lga        ward  population public_meeting              recorded_by  \
0     Ludewa    Mundindi         109           True  GeoData Consultants Ltd   
1  Serengeti       Natta         280            NaN  GeoData Consultants Ltd   
2  Simanjiro     Ngorika         250           True  GeoData Consultants Ltd   
3   Nanyumbu    Nanyumbu          58           True  GeoData Consultants Ltd   
4    Karagwe  Nyakasimbi           0           True  GeoData Consultants Ltd   

  scheme_management                  scheme_name permit  construction_year  \
0               VWC                        Roman  False               1999   
1             Other                          NaN   True               2010   
2               VWC  Nyumba ya mungu pipe scheme   True               2009   
3               VWC                          NaN   True               1986   
4               NaN                          NaN   True                  0   

  extraction_type extraction_type_group extraction_type_class management  \
0         gravity               gravity               gravity        vwc   
1         gravity               gravity               gravity        wug   
2         gravity               gravity               gravity        vwc   
3     submersible           submersible           submersible        vwc   
4         gravity               gravity               gravity      other   

  management_group         payment payment_type water_quality quality_group  \
0       user-group    pay annually     annually          soft          good   
1       user-group       never pay    never pay          soft          good   
2       user-group  pay per bucket   per bucket          soft          good   
3       user-group       never pay    never pay          soft          good   
4            other       never pay    never pay          soft          good   

       quantity quantity_group                source           source_type  \
0        enough         enough                spring                spring   
1  insufficient   insufficient  rainwater harvesting  rainwater harvesting   
2        enough         enough                   dam                   dam   
3           dry            dry           machine dbh              borehole   
4      seasonal       seasonal  rainwater harvesting  rainwater harvesting   

  source_class              waterpoint_type waterpoint_type_group  \
0  groundwater           communal standpipe    communal standpipe   
1      surface           communal standpipe    communal standpipe   
2      surface  communal standpipe multiple    communal standpipe   
3  groundwater  communal standpipe multiple    communal standp

The combined data set has 59400 rows and 41 columns as seen in the cell below. The data types of the respective columns are also seen below

In [82]:
all_df.shape

(59400, 41)

In [83]:
all_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59400 entries, 0 to 59399
Data columns (total 41 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   id                     59400 non-null  int64  
 1   amount_tsh             59400 non-null  float64
 2   date_recorded          59400 non-null  object 
 3   funder                 55763 non-null  object 
 4   gps_height             59400 non-null  int64  
 5   installer              55745 non-null  object 
 6   longitude              59400 non-null  float64
 7   latitude               59400 non-null  float64
 8   wpt_name               59398 non-null  object 
 9   num_private            59400 non-null  int64  
 10  basin                  59400 non-null  object 
 11  subvillage             59029 non-null  object 
 12  region                 59400 non-null  object 
 13  region_code            59400 non-null  int64  
 14  district_code          59400 non-null  int64  
 15  lg

### Data Cleaning and wrangling

Dealing with missing or NaN values

In [84]:
#outputs columns and thier missing value numbers in descending order
all_df.isna().sum().sort_values(ascending=False)

scheme_name              28810
scheme_management         3878
installer                 3655
funder                    3637
public_meeting            3334
permit                    3056
subvillage                 371
wpt_name                     2
water_quality                0
extraction_type_class        0
management                   0
management_group             0
payment                      0
payment_type                 0
quantity                     0
quality_group                0
extraction_type              0
quantity_group               0
source                       0
source_type                  0
source_class                 0
waterpoint_type              0
waterpoint_type_group        0
extraction_type_group        0
id                           0
construction_year            0
amount_tsh                   0
recorded_by                  0
population                   0
ward                         0
lga                          0
district_code                0
region_c

In [85]:
missing_data = all_df.isnull().sum()
missing_data = missing_data.drop(missing_data[missing_data == 0].index).sort_values(ascending=False)
missing_df = pd.DataFrame({'missing number' :missing_data})
missing_df['missing percent'] = (missing_df['missing number'] / len(all_df)) * 100
missing_df


missing number  missing percent
scheme_name                 28810        48.501684
scheme_management            3878         6.528620
installer                    3655         6.153199
funder                       3637         6.122896
public_meeting               3334         5.612795
permit                       3056         5.144781
subvillage                    371         0.624579
wpt_name                        2         0.003367

In [ ]:
missing_data.plot(kind='bar', figsize=(10,6))
plt.ylabel("Number of missing records")
plt.xlabel("Data columns in the data")
plt.title("Graph of missing records in the dataset")

From the cells above we can see that the columns with the most missing values. They are represented with a bar chart showing the number of records missing in he columns. From the cells above we can see that the column `scheme_name` has the most at 28810 records an the rest being around 3000 missing records and finally the las which is `wpt_name` being at around 300. From this we can see that most of these columns store object or string data types

With 48.5% of the `scheme_name` data missing dropping this data column is the right decision. We could try replacing the missing values with a place holder word like "other" but considering that it stores names of the water pumping projects, there are very few reoccuring values in the data column hence this option is not feasible.



For the binary data columns `public_meeting` and `permit` they are going to be replaced by the most likely outcome

In [86]:
#dropping column scheme_name
all_df.drop(['scheme_name'], axis=1, inplace=True)
all_df.shape

(59400, 40)

For the other data features `scheme_management`, `installer`, `funder`, `subvillage`, `wpt_name` it is feasible to replace them with other

In [87]:
#replacing null values in these columns with "other"
missed = ["installer", "funder", "subvillage", "wpt_name", "scheme_management"]
for m in missed:
    all_df[m].fillna("other",inplace=True)

For the binary data columns `public_meeting` and `permit` they are going to be replaced according to a criteria.


In [88]:
#most likely outcome is True
all_df['public_meeting'].value_counts(dropna=False)

#filling null values with True
all_df['public_meeting'].fillna(True, inplace=True)

In the case of `public_meeting` above, the null values were replaced by True. This is mainly because we can say that a verbal meeting could have taken place and there might not have been a physical record. Hence we are going to use a bias that the meeting most likely happened.

For `permit` however, we are going to act with a bias of permit is a legal document and in this case a permit is more likely not to exist that it does exist. This is because a permit is issued by a governing institution and the government records do exist.

In [89]:
all_df['permit'].value_counts(dropna=False)

all_df['permit'].fillna(False, inplace=True)

In [90]:
#rechecking all the null values

all_df.isna().sum().sort_values(ascending=False)

id                       0
amount_tsh               0
construction_year        0
extraction_type          0
extraction_type_group    0
extraction_type_class    0
management               0
management_group         0
payment                  0
payment_type             0
water_quality            0
quality_group            0
quantity                 0
quantity_group           0
source                   0
source_type              0
source_class             0
waterpoint_type          0
waterpoint_type_group    0
permit                   0
scheme_management        0
recorded_by              0
num_private              0
date_recorded            0
funder                   0
gps_height               0
installer                0
longitude                0
latitude                 0
wpt_name                 0
basin                    0
public_meeting           0
subvillage               0
region                   0
region_code              0
district_code            0
lga                      0
w

As you can see from above there are no more null 



### Dealing with 0s in numerical values

This section deals with replacing the 0s in some of the numeric data types in the columns with either the mean or the median values. The numeric data types in the dataset represent various different information. Some of which having a 0 as a value is a possibility. In this section thiugh we are going to focus on the data columns where having 0 as an input is not feasible. These data features include `construction_year` and `population`

Replacing 0s in the `population` data column with the mean of the whole column



In [92]:
#convering 0s to null
all_df['population'] = all_df['population'].replace(0, np.nan)


#replacing null values with the mean
population_avg = all_df['population'].mean(skipna=True)
all_df['population'].fillna(round(population_avg, 0), inplace=True)

count_pop = (all_df['population'] == 0).sum()
all_df['population'].head()

0    109.0
1    280.0
2    250.0
3     58.0
4    281.0
Name: population, dtype: float64

Replacing 0s in the `construction_year` data column with the mean of the whole column

In [94]:
#turning 0s to null values
all_df['construction_year'] = all_df['construction_year'].replace(0, np.nan)

#replacing null in the construction_year with the mean
year_avg = round(all_df['construction_year'].mean(skipna=True), 0)
all_df['construction_year'].fillna(int(year_avg), inplace=True)

count_missing_year = (all_df['construction_year'] == 0).sum()
all_df['construction_year'].head()

0    1999.0
1    2010.0
2    2009.0
3    1986.0
4    1997.0
Name: construction_year, dtype: float64

### Feature engineering

In this section some data columns are going to be altered to make sure they are in a good format and some new data features are going to be created from already existing ones.

In the cell above we are going to convert date_recorder attribute into a date and time format. This will enable us to easily manipulate or use it in calculations further down the process.

In [95]:
#converts date_recoded into date and time format
all_df['date_recorded'] = pd.to_datetime(all_df['date_recorded'])
all_df['date_recorded'].dt.year.head()

0    2011
1    2013
2    2013
3    2013
4    2011
Name: date_recorded, dtype: int32

The age of a waterpoint plays a huge role in determining whether a particular waterpoint is likely to breakdown or need maintenance. This can be because the water pump has been in service for way longer than a new one hence more likely to breakdown or outdated technology from the pumps used in older waterpoint are more unreliable and more likely to breakdown. This is why a feature like this needs to be created and added into the dataset.

In [96]:
#creates a new feature operational_years
all_df['operational_years'] = all_df['date_recorded'].dt.year - all_df['construction_year']
all_df['operational_years'].head(6)

0    12.0
1     3.0
2     4.0
3    27.0
4    14.0
5     2.0
Name: operational_years, dtype: float64

In [97]:
all_df.shape

(59400, 41)

In [ ]:
all_df.drop(['id','recorded_by','date_recorded'], axis = 1, inplace = True)

In [ ]:
plt.figure(figsize=(14,6))
sns.countplot(data=all_df, x='water_quality', hue='status_group')
plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)
all_df.water_quality.value_counts()

In [ ]:
all_df.loc[all_df['amount_tsh'] == 0].shape

In [ ]:
plt.figure(figsize=(24, 9))
sns.countplot(data=all_df,x='region',hue='status_group')

In [ ]:
plt.figure(figsize=(24, 9))
sns.countplot(data=all_df,x='public_meeting',hue='status_group')

In [ ]:
plt.figure(figsize=(24, 9))
sns.FacetGrid(data=all_df, hue='status_group', height=5, aspect=2).map(sns.kdeplot, 'gps_height', shade=True).add_legend()

In [ ]:
# create a list of our conditions
conditions = [
    (all_df['status_group'] == "functional"),
    (all_df['status_group'] == "non functional"),
    (all_df['status_group'] == "functional needs repair")]
# create a list of the values we want to assign for each condition
values = [1, 0, 2]

# create a new column and use np.select to assign values to it using our lists as arguments
all_df['Operation'] = np.select(conditions, values)

In [ ]:
all_df.head()